core/generator/generator.go

```diff
@@ -75,7 +75,16 @@ func Generate(
 		log.Fatal(ctx, log.KeyError, err)
 	}
 	if b != nil && (g.latestBlock == nil || b.Height == g.latestBlock.Height+1) {
-		s := state.Copy(g.latestSnapshot)
+		var s *state.Snapshot
+		if g.latestSnapshot == nil {
```

Comment:
> Can we pass the blockchain ID into Recover instead?

Looking at the code, the best way to do that seems to be by storing it in the `Chain` object as in the earlier version of this PR. I'm fine with that, but maybe you're not?